File: bindings/typescript/sysprims/test/containment.test.ts
import assert from "node:assert/strict";
import { execFileSync } from "node:child_process";
import { setTimeout as delay } from "node:timers/promises";
import test from "node:test";

import fs from "node:fs";
import os from "node:os";
import path from "node:path";

import { loadSysprims } from "../src/ffi";
import { ContainedProcess, spawnContained, SysprimsError, SysprimsErrorCode } from "../src/index";

function isInvalid(error: unknown): boolean {
  return error instanceof SysprimsError && error.code === SysprimsErrorCode.InvalidArgument;
}

test("spawnContained rejects empty argv and signal 0", async () => {
  await assert.rejects(() => spawnContained([]), isInvalid);
  await assert.rejects(() => spawnContained(["true"], { signal: 0 }), isInvalid);
});

test("spawnContained happy path, dispose twice, and stale reuse", async (t) => {
  if (os.platform() === "win32") {
    t.skip("managed spawn is unix-only until Job assignment exists");
    return;
  }

  const first = await spawnContained(["true"], { graceTimeoutMs: 50, killTimeoutMs: 200 });
  const snap = await first.wait({ timeoutMs: 5000 });
  assert.equal(snap.leader_status, "completed");
  assert.equal(snap.tree_kill_reliability, "guaranteed");
  assert.equal(snap.boundary_strength, "cooperative_group");
  assert.ok(snap.completion?.status === "empty" || snap.completion?.status === "unknown");
  await first.close();
  await first.close();
  await assert.rejects(() => first.poll(), isInvalid);

  const second = await spawnContained(["true"], { graceTimeoutMs: 50, killTimeoutMs: 200 });
  await assert.rejects(() => first.identity(), isInvalid);
  await second.close();
});

test("terminate keeps spawn reliability and second call is inert", async (t) => {
  if (os.platform() === "win32") {
    t.skip("managed spawn is unix-only until Job assignment exists");
    return;
  }

  const handle = await spawnContained(["sleep", "30"], {
    graceTimeoutMs: 50,
    killTimeoutMs: 500,
  });
  const before = await handle.identity();
  const snap = await handle.terminate();
  assert.equal(snap.leader_status, "terminated");
  assert.equal(snap.tree_kill_reliability, before.tree_kill_reliability);
  const again = await handle.terminate();
  assert.equal(again.handle_state, "inert");
  await handle.close();
});

test("cancel wait during terminate leaves native owned or inert", async (t) => {
  if (os.platform() === "win32") {
    t.skip("managed spawn is unix-only until Job assignment exists");
    return;
  }

  const handle = await spawnContained(["sleep", "30"], {
    graceTimeoutMs: 80,
    killTimeoutMs: 500,
  });
  const controller = new AbortController();
  const waitPromise = handle.wait({ timeoutMs: 5000, signal: controller.signal });
  const terminatePromise = handle.terminate();
  controller.abort();
  await assert.rejects(
    () => waitPromise,
    (error: unknown) => {
      return error instanceof SysprimsError && error.code === SysprimsErrorCode.Timeout;
    },
  );
  const terminated = await terminatePromise;
  assert.ok(terminated.handle_state === "inert" || terminated.leader_status === "terminated");
  const poll = await handle.poll();
  assert.equal(poll.handle_state, "inert");
  await handle.close();
});

test("close races wait and terminate", async (t) => {
  if (os.platform() === "win32") {
    t.skip("managed spawn is unix-only until Job assignment exists");
    return;
  }

  const handle = await spawnContained(["sleep", "30"], {
    graceTimeoutMs: 50,
    killTimeoutMs: 500,
  });
  const [waitResult, terminateResult, closeResult] = await Promise.allSettled([
    handle.wait({ timeoutMs: 2000 }),
    handle.terminate(),
    handle.close(),
  ]);
  if (closeResult.status === "rejected") {
    await handle.identity();
    await handle.close();
  } else {
    await handle.close();
    await assert.rejects(() => handle.identity(), isInvalid);
  }
  for (const result of [waitResult, terminateResult]) {
    if (result.status === "rejected") {
      assert.ok(isInvalid(result.reason), `unexpected race error: ${result.reason}`);
    } else {
      assert.equal(result.value.handle_state, "inert");
      assert.equal(result.value.leader_status, "terminated");
    }
  }
  if (waitResult.status === "fulfilled" && terminateResult.status === "fulfilled") {
    assert.deepEqual(waitResult.value.identity, terminateResult.value.identity);
    assert.deepEqual(waitResult.value.completion, terminateResult.value.completion);
  }
});

test("failed close keeps the token retryable", async (t) => {
  if (os.platform() === "win32") {
    t.skip("managed spawn is unix-only until Job assignment exists");
    return;
  }

  const handle = await spawnContained(["sleep", "30"], {
    graceTimeoutMs: 50,
    killTimeoutMs: 500,
  });
  const native = loadSysprims();
  const original = native.sysprimsContainmentClose;
  native.sysprimsContainmentClose = async () => ({
    code: SysprimsErrorCode.InvalidArgument,
    message: "forced close failure",
  });
  try {
    await assert.rejects(() => handle.close(), isInvalid);
    await handle.identity();
  } finally {
    native.sysprimsContainmentClose = original;
  }
  await handle.close();
  await assert.rejects(() => handle.identity(), isInvalid);
});

test("already-aborted wait does not launch after abort", async (t) => {
  if (os.platform() === "win32") {
    t.skip("managed spawn is unix-only until Job assignment exists");
    return;
  }

  const handle = await spawnContained(["true"], { graceTimeoutMs: 50, killTimeoutMs: 200 });
  const controller = new AbortController();
  controller.abort();
  await assert.rejects(
    () => handle.wait({ timeoutMs: 5000, signal: controller.signal }),
    (error: unknown) => error instanceof SysprimsError && error.code === SysprimsErrorCode.Timeout,
  );
  await handle.close();
});

test("invalid high signal does not spawn", async () => {
  const marker = path.join(os.tmpdir(), `sysprims-ts-high-signal-${process.pid}.marker`);
  try {
    fs.rmSync(marker, { force: true });
  } catch {
    // absent is fine
  }
  await assert.rejects(() => spawnContained(["touch", marker], { signal: 99 }), isInvalid);
  assert.equal(fs.existsSync(marker), false);
});

test("windows managed spawn fails before spawn", async (t) => {
  if (os.platform() !== "win32") {
    t.skip("windows-only rejection fixture");
    return;
  }
  const marker = path.join(fs.mkdtempSync(path.join(os.tmpdir(), "sysprims-no-spawn-")), "marker");
  t.after(() => fs.rmSync(path.dirname(marker), { recursive: true, force: true }));
  await assert.rejects(
    () => spawnContained(["cmd", "/C", `echo spawned>"${marker}"`]),
    (error: unknown) =>
      error instanceof SysprimsError && error.code === SysprimsErrorCode.NotSupported,
  );
  assert.equal(fs.existsSync(marker), false);
});

test("ContainedProcess is a class", () => {
  assert.equal(typeof ContainedProcess, "function");
});

for (const [name, argv, expected] of [
  ["long-lived", ["sleep", "30"], "timed_out"],
  ["fast", ["true"], "completed"],
] as const) {
  test(`native deadline with delayed first observation: ${name}`, async (t) => {
    if (os.platform() === "win32") {
      t.skip("Unix containment");
      return;
    }
    const handle = await spawnContained([...argv], {
      executionTimeoutMs: 200,
      graceTimeoutMs: 20,
      killTimeoutMs: 200,
    });
    try {
      await delay(450);
      const deadline = performance.now() + 3000;
      let observed = await handle.identity();
      while (observed.handle_state !== "inert" && performance.now() < deadline) {
        await delay(10);
        observed = await handle.identity();
      }
      assert.equal(observed.handle_state, "inert");
      const result = await handle.wait({ timeoutMs: 1000 });
      assert.equal(result.leader_status, expected);
      if (expected === "completed") assert.equal(result.timed_out, false);
    } finally {
      await handle.close();
    }
  });
}

test("bounded wait returns while another operation finalizes", async (t) => {
  if (os.platform() === "win32") {
    t.skip("Unix containment");
    return;
  }
  const handle = await spawnContained(["sleep", "30"], { graceTimeoutMs: 800, killTimeoutMs: 200 });
  const terminating = handle.terminate();
  try {
    await delay(100);
    const started = performance.now();
    const result = await handle.wait({ timeoutMs: 10 });
    const elapsed = performance.now() - started;
    assert.equal(result.handle_state, "active");
    assert.equal(result.leader_status, "running");
    assert.ok(elapsed < 300, `10ms wait took ${elapsed}ms`);
    assert.equal((await terminating).leader_status, "terminated");
  } finally {
    await terminating;
    await handle.close();
  }
});

test("explicit close unregisters finalizer and never closes twice", (t) => {
  if (os.platform() === "win32") {
    t.skip("Unix containment");
    return;
  }
  // Isolate a deterministic FinalizationRegistry scheduler from the suite.
  // Flushing it represents a GC finalization turn after explicit close.
  execFileSync(
    process.execPath,
    [
      "-e",
      `
    const assert = require('node:assert/strict');
    const registries = [];
    global.FinalizationRegistry = class {
      constructor(callback) { this.callback = callback; this.entries = new Map(); registries.push(this); }
      register(target, value, token) { this.entries.set(token, value); }
      unregister(token) { return this.entries.delete(token); }
      flush() { for (const value of this.entries.values()) this.callback(value); this.entries.clear(); }
    };
    const {spawnContained} = require(${JSON.stringify(path.resolve(__dirname, "../src/index.js"))});
    const {loadSysprims} = require(${JSON.stringify(path.resolve(__dirname, "../src/ffi.js"))});
    (async () => {
      const native = loadSysprims();
      const close = native.sysprimsContainmentClose;
      let calls = 0;
      native.sysprimsContainmentClose = (...args) => { calls++; return close(...args); };
      const handle = await spawnContained(['true'], {graceTimeoutMs:20, killTimeoutMs:200});
      assert.equal(registries.length, 1);
      assert.equal(registries[0].entries.size, 1);
      await handle.close();
      await handle.close();
      assert.equal(registries[0].entries.size, 0);
      registries[0].flush();
      await new Promise(resolve => setImmediate(resolve));
      assert.equal(calls, 1);
    })().catch(error => { console.error(error); process.exitCode = 1; });
  `,
    ],
    { timeout: 10000, stdio: "pipe" },
  );
});

File: bindings/go/sysprims/containment_test.go
package sysprims_test

import (
	"os"
	"path/filepath"
	"runtime"
	"sync"
	"sync/atomic"
	"testing"
	"time"

	"github.com/3leaps/sysprims/bindings/go/sysprims"
)

func spawnTrue(t *testing.T) *sysprims.ContainedProcess {
	t.Helper()
	handle, err := sysprims.SpawnContained(sysprims.SpawnContainedConfig{
		Argv:           []string{"true"},
		GraceTimeoutMS: uint64Ptr(50),
		KillTimeoutMS:  uint64Ptr(200),
	})
	if err != nil {
		t.Fatalf("SpawnContained(true): %v", err)
	}
	return handle
}

func spawnSleep(t *testing.T) *sysprims.ContainedProcess {
	t.Helper()
	handle, err := sysprims.SpawnContained(sysprims.SpawnContainedConfig{
		Argv:           []string{"sleep", "30"},
		GraceTimeoutMS: uint64Ptr(50),
		KillTimeoutMS:  uint64Ptr(500),
	})
	if err != nil {
		t.Fatalf("SpawnContained(sleep): %v", err)
	}
	return handle
}

func uint64Ptr(v uint64) *uint64 { return &v }

func TestSpawnContainedRejectsEmptyArgvAndSignalZero(t *testing.T) {
	_, err := sysprims.SpawnContained(sysprims.SpawnContainedConfig{})
	if err == nil {
		t.Fatal("empty argv should fail")
	}
	zero := int32(0)
	_, err = sysprims.SpawnContained(sysprims.SpawnContainedConfig{
		Argv:   []string{"true"},
		Signal: &zero,
	})
	if err == nil {
		t.Fatal("signal 0 should fail")
	}
}

func TestSpawnContainedInvalidForeignTokensFailClosed(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("managed spawn is unix-only until Job assignment exists")
	}
	handle := spawnTrue(t)
	if err := handle.Close(); err != nil {
		t.Fatal(err)
	}
	if err := handle.Close(); err != nil {
		t.Fatalf("language dispose-twice must be a no-op: %v", err)
	}
	if _, err := handle.Identity(); err == nil {
		t.Fatal("closed handle identity must fail closed")
	}
}

func TestSpawnContainedHappyPathAndStaleReuse(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("managed spawn is unix-only until Job assignment exists")
	}
	first := spawnTrue(t)
	snap, err := first.Wait(5 * time.Second)
	if err != nil {
		t.Fatal(err)
	}
	if snap.LeaderStatus != "completed" {
		t.Fatalf("leader_status=%q", snap.LeaderStatus)
	}
	if snap.TreeKillReliability != "guaranteed" {
		t.Fatalf("reliability=%q", snap.TreeKillReliability)
	}
	if snap.BoundaryStrength != "cooperative_group" {
		t.Fatalf("boundary=%q", snap.BoundaryStrength)
	}
	if snap.Completion == nil || (snap.Completion.Status != "empty" && snap.Completion.Status != "unknown") {
		t.Fatalf("completion=%+v", snap.Completion)
	}
	if err := first.Close(); err != nil {
		t.Fatal(err)
	}

	second := spawnTrue(t)
	if _, err := first.Poll(); err == nil {
		t.Fatal("stale token after slot reuse must fail closed")
	}
	if err := second.Close(); err != nil {
		t.Fatal(err)
	}
}

func TestSpawnContainedTerminateKeepsSpawnReliability(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("managed spawn is unix-only until Job assignment exists")
	}
	handle := spawnSleep(t)
	defer handle.Close()
	before, err := handle.Identity()
	if err != nil {
		t.Fatal(err)
	}
	snap, err := handle.Terminate()
	if err != nil {
		t.Fatal(err)
	}
	if snap.LeaderStatus != "terminated" {
		t.Fatalf("leader_status=%q", snap.LeaderStatus)
	}
	if snap.TreeKillReliability != before.TreeKillReliability {
		t.Fatalf("reliability mutated after terminate")
	}
	again, err := handle.Terminate()
	if err != nil {
		t.Fatal(err)
	}
	if again.HandleState != "inert" {
		t.Fatalf("second terminate should be inert, got %q", again.HandleState)
	}
}

func TestSpawnContainedCloseRacesWaitAndTerminate(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("managed spawn is unix-only until Job assignment exists")
	}
	handle := spawnSleep(t)
	var wg sync.WaitGroup
	var closeErr, waitErr, termErr error
	var waited, terminated *sysprims.ContainmentSnapshot
	wg.Add(3)
	go func() {
		defer wg.Done()
		waited, waitErr = handle.Wait(2 * time.Second)
	}()
	go func() {
		defer wg.Done()
		terminated, termErr = handle.Terminate()
	}()
	go func() {
		defer wg.Done()
		time.Sleep(20 * time.Millisecond)
		closeErr = handle.Close()
	}()
	wg.Wait()
	for name, result := range map[string]struct {
		snapshot *sysprims.ContainmentSnapshot
		err      error
	}{
		"wait": {waited, waitErr}, "terminate": {terminated, termErr},
	} {
		if result.err != nil {
			nativeErr, ok := result.err.(*sysprims.Error)
			if !ok || nativeErr.Code != sysprims.ErrInvalidArgument {
				t.Fatalf("%s unexpected error: %v", name, result.err)
			}
		} else if result.snapshot == nil || result.snapshot.HandleState != "inert" || result.snapshot.LeaderStatus != "terminated" {
			t.Fatalf("%s unexpected successful outcome: %+v", name, result.snapshot)
		}
	}
	if waited != nil && terminated != nil && waited.LeaderStatus != terminated.LeaderStatus {
		t.Fatal("racing methods returned different terminal outcomes")
	}

	if closeErr != nil {
		if _, err := handle.Identity(); err != nil {
			t.Fatalf("failed close must preserve owner (wait=%v terminate=%v identity=%v)", waitErr, termErr, err)
		}
		if err := handle.Close(); err != nil {
			t.Fatalf("retry close after failure: %v", err)
		}
	}
	if err := handle.Close(); err != nil {
		t.Fatalf("final close: %v", err)
	}
	if _, err := handle.Identity(); err == nil {
		t.Fatal("closed handle identity must fail")
	}
}

func TestSpawnContainedWaitRejectsNegativeAndKeepsSubMillisecondFinite(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("managed spawn is unix-only until Job assignment exists")
	}
	handle := spawnSleep(t)
	defer handle.Close()
	if _, err := handle.Wait(-time.Second); err == nil {
		t.Fatal("negative wait must fail")
	}
	started := time.Now()
	if _, err := handle.Wait(500 * time.Microsecond); err != nil {
		t.Fatalf("positive sub-ms wait must stay finite: %v", err)
	}
	if time.Since(started) > 3*time.Second {
		t.Fatal("positive sub-ms wait was treated as infinite")
	}
	if _, err := handle.Terminate(); err != nil {
		t.Fatalf("terminate after sub-ms wait: %v", err)
	}
}

func TestSpawnContainedCloseFailureRestoresToken(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("managed spawn is unix-only until Job assignment exists")
	}
	handle := spawnSleep(t)
	sysprims.SetNativeCloseForTest(func(uint64) error {
		return &sysprims.Error{Code: sysprims.ErrInvalidArgument, Message: "forced close failure"}
	})
	err := handle.Close()
	sysprims.SetNativeCloseForTest(nil)
	if err == nil {
		t.Fatal("forced native close failure must surface")
	}
	if _, idErr := handle.Identity(); idErr != nil {
		_ = handle.Close()
		t.Fatalf("failed close must keep the token usable: %v", idErr)
	}
	if err := handle.Close(); err != nil {
		t.Fatalf("retry close: %v", err)
	}
}

func TestSpawnContainedInvalidHighSignalDoesNotSpawn(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("managed spawn is unix-only until Job assignment exists")
	}
	marker := t.TempDir() + "/should-not-exist"
	sig := int32(99)
	_, err := sysprims.SpawnContained(sysprims.SpawnContainedConfig{
		Argv:   []string{"touch", marker},
		Signal: &sig,
	})
	if err == nil {
		t.Fatal("signal 99 must fail before spawn")
	}
	if _, statErr := os.Stat(marker); statErr == nil {
		t.Fatal("invalid high signal must not start argv")
	}
}

func TestSpawnContainedWindowsUnsupported(t *testing.T) {
	if runtime.GOOS != "windows" {
		t.Skip("windows-only rejection fixture")
	}
	marker := filepath.Join(t.TempDir(), "should-not-exist")
	_, err := sysprims.SpawnContained(sysprims.SpawnContainedConfig{
		Argv: []string{"cmd", "/C", "echo spawned>\"" + marker + "\""},
	})
	if _, statErr := os.Stat(marker); !os.IsNotExist(statErr) {
		t.Fatalf("spawn marker: %v", statErr)
	}
	if err == nil {
		t.Fatal("windows managed spawn must fail before spawn")
	}
	sErr, ok := err.(*sysprims.Error)
	if !ok || sErr.Code != sysprims.ErrNotSupported {
		t.Fatalf("expected ErrNotSupported, got %v", err)
	}
}

func TestContainedNativeExecutionDeadline(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("Unix containment")
	}
	for _, test := range []struct {
		name   string
		argv   []string
		status string
	}{
		{"running", []string{"sleep", "30"}, "timed_out"},
		{"fast", []string{"true"}, "completed"},
	} {
		t.Run(test.name, func(t *testing.T) {
			h, err := sysprims.SpawnContained(sysprims.SpawnContainedConfig{Argv: test.argv, ExecutionTimeoutMS: uint64Ptr(200), GraceTimeoutMS: uint64Ptr(20), KillTimeoutMS: uint64Ptr(200)})
			if err != nil {
				t.Fatal(err)
			}
			defer h.Close()
			time.Sleep(450 * time.Millisecond)
			observed, err := h.Identity()
			deadline := time.Now().Add(3 * time.Second)
			for err == nil && observed.HandleState != "inert" && time.Now().Before(deadline) {
				time.Sleep(10 * time.Millisecond)
				observed, err = h.Identity()
			}
			if err != nil || observed.HandleState != "inert" {
				t.Fatalf("native monitor did not finalize: %+v %v", observed, err)
			}
			result, err := h.Wait(time.Second)
			if err != nil || result.LeaderStatus != test.status {
				t.Fatalf("delayed wait: %+v %v", result, err)
			}
			if test.status == "completed" && (result.TimedOut == nil || *result.TimedOut) {
				t.Fatal("natural completion reported cleanup timeout")
			}
		})
	}
}

func TestContainedBoundedWaitDuringTerminate(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("Unix containment")
	}
	h, err := sysprims.SpawnContained(sysprims.SpawnContainedConfig{Argv: []string{"sleep", "30"}, GraceTimeoutMS: uint64Ptr(800), KillTimeoutMS: uint64Ptr(200)})
	if err != nil {
		t.Fatal(err)
	}
	defer h.Close()
	done := make(chan error, 1)
	go func() { _, err := h.Terminate(); done <- err }()
	time.Sleep(100 * time.Millisecond)
	started := time.Now()
	result, waitErr := h.Wait(10 * time.Millisecond)
	elapsed := time.Since(started)
	termErr := <-done
	if termErr != nil {
		t.Fatal(termErr)
	}
	if waitErr != nil || result.HandleState != "active" || result.LeaderStatus != "running" {
		t.Fatalf("wait: %+v %v", result, waitErr)
	}
	if elapsed > 300*time.Millisecond {
		t.Fatalf("10ms wait took %v", elapsed)
	}
}

func TestContainedFinalizerAfterExplicitCloseDoesNotCloseAgain(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("Unix containment")
	}
	h := spawnTrue(t)
	if err := h.Close(); err != nil {
		t.Fatal(err)
	}
	calls := 0
	sysprims.SetNativeCloseForTest(func(uint64) error { calls++; return nil })
	defer sysprims.SetNativeCloseForTest(nil)
	sysprims.RunContainedFinalizerForTest(h)
	if err := h.Close(); err != nil {
		t.Fatal(err)
	}
	if calls != 0 {
		t.Fatalf("native close called %d times after disposal", calls)
	}
}

//go:noinline
func callWithLastReference(t *testing.T, method string, finalized *atomic.Bool) error {
	h := spawnSleep(t)
	sysprims.ObserveFinalizerForTest(h, func() { finalized.Store(true) })
	var err error
	switch method {
	case "identity":
		_, err = h.Identity()
	case "poll":
		_, err = h.Poll()
	case "wait":
		_, err = h.Wait(time.Millisecond)
	case "terminate":
		_, err = h.Terminate()
	}
	return err
}

func TestContainedMethodsKeepReceiverAlive(t *testing.T) {
	if runtime.GOOS == "windows" {
		t.Skip("Unix containment")
	}
	for _, method := range []string{"identity", "poll", "wait", "terminate"} {
		t.Run(method, func(t *testing.T) {
			var finalized atomic.Bool
			var premature bool
			sysprims.SetSnapshotHookForTest(func() {
				for i := 0; i < 5; i++ {
					runtime.GC()
					time.Sleep(10 * time.Millisecond)
				}
				premature = finalized.Load()
			})
			err := callWithLastReference(t, method, &finalized)
			sysprims.SetSnapshotHookForTest(nil)
			deadline := time.Now().Add(3 * time.Second)
			for !finalized.Load() && time.Now().Before(deadline) {
				runtime.GC()
				time.Sleep(10 * time.Millisecond)
			}
			if premature {
				t.Fatal("receiver finalized during native method")
			}
			if err != nil {
				t.Fatal(err)
			}
			if !finalized.Load() {
				t.Fatal("fixture did not become collectible after method")
			}
		})
	}
}
